From 1bdb152c76996e8c5516dcc8607ffb26162ea30c Mon Sep 17 00:00:00 2001
From: Jan Sondhauss <jan.sondhauss@wago.com>
Date: Tue, 4 Sep 2018 09:50:33 +0200
Subject: [PATCH] net: pnsocket: fix not moving tx-consumer-point when
 tx-packet is dropped

This prevents the same buffer from being used multiple times.

Signed-off-by: Jan Sondhauss <jan.sondhauss@wago.com>
---
 net/pn/pnsk.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/net/pn/pnsk.c b/net/pn/pnsk.c
index ab4bb67..15f02cb 100644
--- a/net/pn/pnsk.c
+++ b/net/pn/pnsk.c
@@ -210,6 +210,7 @@ static int pnsk_generic_xmit(struct sock *sk, struct msghdr *m,
 		skb->destructor = pnsk_destruct_skb;
 
 		err = dev_direct_xmit(skb, ps->queue_id);
+		pnskq_discard_desc(ps->tx);
 		/* Ignore NET_XMIT_CN as packet might have been sent */
 		if (err == NET_XMIT_DROP || err == NETDEV_TX_BUSY) {
 			err = -EAGAIN;
@@ -218,7 +219,6 @@ static int pnsk_generic_xmit(struct sock *sk, struct msghdr *m,
 		}
 
 		sent_frame = true;
-		pnskq_discard_desc(ps->tx);
 	}
 
 out:
-- 
2.7.4

